platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/migration/ImportMigrationManagerImpl.java

```diff
@@ -0,0 +1,194 @@
+     *                                  or <code>migratables</code> is <code>null</code>
+     */
+    public ImportMigrationManagerImpl(MigrationReport report, Path exportFile,
+            Stream<? extends Migratable> migratables) {
+        this(report, exportFile, migratables, ImportMigrationManagerImpl.newZipFileFor(exportFile));
+    }
+
+    ImportMigrationManagerImpl(MigrationReport report, Path exportFile,
+            Stream<? extends Migratable> migratables, ZipFile zip) {
+        Validate.notNull(report, "invalid null report");
+        Validate.isTrue(report.getOperation() == MigrationOperation.IMPORT,
+                "invalid migration operation");
+        Validate.notNull(migratables, "invalid null migratables");
+        this.report = report;
+        this.exportFile = exportFile;
+        this.zip = zip;
+        Map<String, Object> metadata;
+
+        try {
+            // pre-create contexts for all registered migratables
+            this.contexts = migratables.collect(Collectors.toMap(Migratable::getId,
+                    m -> new ImportMigrationContextImpl(report, zip, m),
+                    ConfigurationMigrationManager.throwingMerger(),
+                    LinkedHashMap::new)); // to preserved ranking order and remove duplicates
+            // add a system contexts
+            contexts.put(null, new ImportMigrationContextImpl(report, zip));
+            zip.stream()
+                    .map(ze -> new ImportMigrationEntryImpl(this::getContextFor, ze))
+                    .forEach(me -> me.getContext()
+                            .addEntry(me));
+            metadata = retrieveMetadata();
+        } catch (IOException e) {
+            throw new MigrationException(Messages.IMPORT_FILE_READ_ERROR, exportFile, e);
+        }
+        this.version = JsonUtils.getStringFrom(metadata,
+                MigrationContextImpl.METADATA_VERSION,
+                true);
+        if (!MigrationContextImpl.VERSION.equals(version)) {
+            throw new MigrationException(Messages.IMPORT_UNSUPPORTED_VERSION_ERROR,
+                    version,
+                    MigrationContextImpl.VERSION);
+        }
+        this.productVersion = JsonUtils.getStringFrom(metadata,
+                MigrationContextImpl.METADATA_PRODUCT_VERSION,
+                true);
+        // process migratables' metadata
+        JsonUtils.getMapFrom(metadata, MigrationContextImpl.METADATA_MIGRATABLES)
+                .forEach((id, o) -> getContextFor(id).processMetadata(JsonUtils.convertToMap(o)));
+    }
+
+    private static ZipFile newZipFileFor(Path exportFile) {
+        Validate.notNull(exportFile, "invalid null export file");
+        try {
+            return new ZipFile(exportFile.toFile());
+        } catch (FileNotFoundException e) {
+            throw new MigrationException(Messages.IMPORT_FILE_MISSING_ERROR, exportFile, e);
+        } catch (IOException e) {
```

Comment:
❓ do we need to close the zip if we get an IOException?

Reply:
Not here as it would not have been opened anyway.